# Content List 
#### 01. Importing Libraries 
#### 02. Importing Datasets 
#### 03. Mixed-Type Data 
#### 04. Missing Values 
#### 05. Duplicates 
#### 06. Task Submission

# 01. Importing Libraries

In [1]:
#Import libraries
import pandas as pd 
import numpy as np
import os 

# 02. Importing Datasets

In [2]:
path = r"C:\Users\maple\OneDrive\Desktop\CareerFoundry\Instacart Basket Analysis"

In [3]:
#Importing original products dataset 
df_prods = pd.read_csv(os.path.join(path,'02 Data','Original Data', 'products.csv'), index_col = False)

In [4]:
#Importing prepared orders dataset 
df_ords = pd.read_csv(os.path.join(path,'02 Data','Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03. Mixed-Type Data

In [5]:
#Create a dataframe
df_test = pd.DataFrame()

In [6]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
#Check for mixed types 
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
#create consistent datatype in mix column 
df_test['mix'] = df_test['mix'].astype('str')

# 04. Missing Values 

In [10]:
#Find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
#create new dataframe of missing values 
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [12]:
df_nan 

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
#Shape of current dataframe including missing values
df_prods.shape

(49693, 5)

In [14]:
#Create dataframe without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
#Shape of new dataframe without missing values 
df_prods_clean.shape

(49677, 5)

# 05. Duplicates 

In [16]:
#Locating duplicate rows in df_prods_clean and assigning to new dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [18]:
#Shape of df_prods_clean dataframe with duplicate rows 
df_prods_clean.shape 

(49677, 5)

In [19]:
#Create new dataframe without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
#Shape of new dataframe without duplicate rows 
df_prods_clean_no_dups.shape

(49672, 5)

In [21]:
#Check for mixed-type data in your df_prods_clean dataframe.
for col in df_prods_clean.columns.tolist():
  weird = (df_prods_clean[[col]].applymap(type) != df_prods_clean[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods_clean[weird]) > 0:
    print (col)

In [22]:
#Exporting the clean products dataframe without missing or duplicate observations  
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# 06. Task Submission

In [23]:
#Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### When looking at the descriptive stats of the products dataframe, I noticed that the max price is 99,999. This should be investigated further as I doubt any grocery store products are this expensive. 

In [24]:
#Check for mixed-type data in your df_ords dataframe.
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### According to results, there are no mixed-type data

In [25]:
#Checking info on orders dataframe to confirm no mixed data types
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   Unnamed: 0               int64  
 1   order_id                 int64  
 2   user_id                  int64  
 3   number_of_orders_placed  int64  
 4   orders_day_of_week       int64  
 5   order_hour_of_day        int64  
 6   days_since_prior_order   float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


In [26]:
#Run a check for missing values in your df_ords dataframe.
df_ords.isnull().sum()

Unnamed: 0                      0
order_id                        0
user_id                         0
number_of_orders_placed         0
orders_day_of_week              0
order_hour_of_day               0
days_since_prior_order     206209
dtype: int64

In [27]:
#Checking to see if there are any values of 0 in the days_since_prior_order column 
(df_ords['days_since_prior_order']== 0).sum()

67755

In [28]:
#Creating a new dataframe of observations with missing values in order to check to see if the number_of_orders_placed values for the observations are 1, indicating no prior orders.
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [29]:
#Checking the dataframe to confirm the number_of_orders placed values
df_ords_nan.head()
df_ords_nan['number_of_orders_placed'].value_counts

<bound method IndexOpsMixin.value_counts of 0          1
11         1
26         1
39         1
45         1
          ..
3420930    1
3420934    1
3421002    1
3421019    1
3421069    1
Name: number_of_orders_placed, Length: 206209, dtype: int64>

#### The only column with missing values in the orders dataframe is days_since_prior_order.  My first hypothesis was that these missing values could either mean that there have been 0 days since the prior order, or that there have been no prior orders. After finding a count of 67,755 observations with a value of 0 in the column, I ruled out that the missing values mean there are 0 days since the prior order, and my leading hypothesis are that these missing values indicate that no prior orders have been placed by the associated user. This was confirmed by the number_of_orders_placed equaling 1 for the records. 

In [30]:
#Address the missing values using an appropriate method.
#Creating a new dataframe with additional column with boolean value identifying missing days_since_prior_order
df_ords_clean = df_ords
df_ords_clean['no_prior_orders'] = df_ords['number_of_orders_placed'] == 1

In [31]:
df_ords_clean

Unnamed: 0  order_id  user_id  number_of_orders_placed  \
0                 0   2539329        1                        1   
1                 1   2398795        1                        2   
2                 2    473747        1                        3   
3                 3   2254736        1                        4   
4                 4    431534        1                        5   
...             ...       ...      ...                      ...   
3421078     3421078   2266710   206209                       10   
3421079     3421079   1854736   206209                       11   
3421080     3421080    626363   206209                       12   
3421081     3421081   2977660   206209                       13   
3421082     3421082    272231   206209                       14   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                         2                  8                     NaN   
1                         3                  7                    15.0   
2                         3                 12                    21.0   
3                         4                  7                    29.0   
4                         4                 15                    28.0   
...                     ...                ...                     ...   
3421078                   5                 18                    29.0   
3421079                   4                 10                    30.0   
3421080                   1                 12                    18.0   
3421081                   1                 12                     7.0   
3421082                   6                 14                    30.0   

         no_prior_orders  
0                   True  
1                  False  
2                  False  
3                  False  
4                  False  
...                  ...  
3421078            False  
3421079            False  
3421080            False  
3421081            False  
3421082            False  

[3421083 rows x 8 columns]

In [36]:
df_ords_clean['no_prior_orders'].value_counts() 

False    3214874
True      206209
Name: no_prior_orders, dtype: int64

#### I chose to handle the missing values by creating a new column with boolean datatype for identifying if a customer has had no prior orders. By doing so I didn't remove the missing values or impute them, as they are accurately missing for customers who have only placed one order. 

In [32]:
#Run a check for duplicate values in your df_ords data.
#Locating duplicate rows in df_ords_clean and assigning to new dataframe
df_ords_clean_dups = df_ords_clean[df_ords_clean.duplicated()]

In [33]:
df_ords_clean_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, number_of_orders_placed, orders_day_of_week, order_hour_of_day, days_since_prior_order, no_prior_orders]
Index: []

In [34]:
df_ords_clean.shape

(3421083, 8)

#### There are no duplicates in the df_ords_clean dataframe 

In [35]:
#Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

#Products dataframe already exported as part of exercise. 